app/scripts/controllers/network/network-controller.test.js

```diff
@@ -50,224 +69,4198 @@ const PRE_1559_BLOCK = {
+      await withController(
+        {
+          state: {
+            networkDetails: {
+              EIPS: {},
             },
+          },
+        },
+        async ({ controller, network }) => {
+          network.mockEssentialRpcCalls({
+            latestBlock: POST_1559_BLOCK,
           });
           await controller.initializeProvider();
+
           await controller.getEIP1559Compatibility();
-          expect(controller.networkDetails.getState().EIPS[1559]).toStrictEqual(
+
+          expect(controller.store.getState().networkDetails.EIPS[1559]).toBe(
             true,
           );
-          await new Promise((resolve) => {
-            controller.on(NETWORK_EVENTS.NETWORK_DID_CHANGE, () => {
-              resolve();
-            });
-            controller.setProviderType('mainnet');
+        },
+      );
+    });
+
+    it('returns false when baseFeePerGas is not in the block header', async () => {
+      await withController(async ({ controller, network }) => {
+        network.mockEssentialRpcCalls({
+          latestBlock: PRE_1559_BLOCK,
+        });
+        await controller.initializeProvider();
+
+        const supportsEIP1559 = await controller.getEIP1559Compatibility();
+
+        expect(supportsEIP1559).toBe(false);
+      });
+    });
+
+    it('persists to state that the network does not support EIP-1559 when baseFeePerGas is not in the block header', async () => {
+      await withController(
+        {
+          state: {
+            networkDetails: {
+              EIPS: {},
+            },
+          },
+        },
+        async ({ controller, network }) => {
+          network.mockEssentialRpcCalls({
+            latestBlock: PRE_1559_BLOCK,
           });
-          expect(
-            controller.networkDetails.getState().EIPS[1559],
-          ).toBeUndefined();
+          await controller.initializeProvider();
+
           await controller.getEIP1559Compatibility();
-          expect(controller.networkDetails.getState().EIPS[1559]).toStrictEqual(
+
+          expect(controller.store.getState().networkDetails.EIPS[1559]).toBe(
             false,
           );
+        },
+      );
+    });
+
+    it('does not make multiple requests to eth_getBlockByNumber when called multiple times (as long as the request to eth_getBlockByNumber succeeded the first time)', async () => {
+      await withController(async ({ controller, network }) => {
+        // This mocks eth_getBlockByNumber once by default
+        network.mockEssentialRpcCalls();
+        await withoutCallingGetEIP1559Compatibility({
+          controller,
+          operation: async () => {
+            await controller.initializeProvider();
+          },
         });
+
+        await controller.getEIP1559Compatibility();
+        await controller.getEIP1559Compatibility();
+
+        expect(network.nockScope.isDone()).toBe(true);
       });
     });
   });
-});
+
+  describe('lookupNetwork', () => {
```

Comment:
Are you talking about this behavior? https://github.com/MetaMask/metamask-extension/blob/884e7022a6e50030fefd1df52db7b99991b15952/app/scripts/controllers/network/network-controller.js#L428-L431

Reply:
Sorry no, I explained that poorly. I mean the behavior covered by the test "emits infuraIsUnblocked as long as the network has not changed by the time the request ends"

If the network changed by the time the request ends, then no state changes happen (e.g. network state, network details, EIP1559 compatibility). Due to this line: https://github.com/MetaMask/metamask-extension/blob/02b2d7fa42e5edd7d66c28a40c5a2f30cc27c1ac/app/scripts/controllers/network/network-controller.js#L217

Reply:
Oh I see, that test isn't testing that condition, but it's a similar one (within `_checkInfuraAvailability`). That one is covered, but this one in `lookupNetwork` is not.

Reply:
Gotcha. Yes that's true.

That case is a bit tricky to test because we have to end up stubbing `ethQuery.sendAsync` so that it switches the network before resolving (because the network is never hit in the Infura case). I'll work on this more tomorrow.

Reply:
Ah right, hadn't considered that this wouldn't apply for the Infura case. Maybe we could just test this for custom networks in that case. Seems fine to ignore this for Infura if the situation isn't possible.

Reply:
Okay, it took a bit for the Infura case, but I think I got this figured out: [b304549](https://github.com/MetaMask/metamask-extension/pull/17498/commits/b30454990d7f20f84847672623edd48836959baa)